File: theory/src/main/kotlin/app/guitar/theory/RhythmPhrases.kt
package app.guitar.theory

/** One click of a phrase: its global 16th-slot index and whether it's a bar-downbeat
 *  accent. */
data class PhraseOnset(val slot: Int, val accent: Boolean)

/**
 * A multi-bar rhythmic phrase: [beats] one-beat units (each subdivision 4) laid out
 * as [bars] bars of [beatsPerBar] beats (time signature beatsPerBar/4). Because every
 * beat is a 16th-grid unit, the whole phrase maps onto a clean 16th grid.
 */
data class RhythmPhrase(
    val bars: Int,
    val beatsPerBar: Int,
    val beats: List<RhythmUnit>,
) {
    init {
        require(beats.size == bars * beatsPerBar) { "phrase must have bars*beatsPerBar beats" }
        require(beats.all { it.subdivision == RhythmPhrases.SLOTS_PER_BEAT }) {
            "phrase beats must be 16th-grid units (subdivision ${RhythmPhrases.SLOTS_PER_BEAT})"
        }
    }

    val slotsPerBeat: Int get() = RhythmPhrases.SLOTS_PER_BEAT
    val totalSlots: Int get() = beats.size * slotsPerBeat
    val slotsPerBar: Int get() = beatsPerBar * slotsPerBeat

    /** Global 16th-slot of every click, accented on bar downbeats that carry an onset. */
    fun onsets(): List<PhraseOnset> {
        val out = ArrayList<PhraseOnset>()
        beats.forEachIndexed { b, unit ->
            val isBarStart = b % beatsPerBar == 0
            for (f in unit.clickFractions()) {
                val local = (f * slotsPerBeat).toInt()
                out.add(PhraseOnset(b * slotsPerBeat + local, accent = isBarStart && local == 0))
            }
        }
        return out
    }
}

object RhythmPhrases {
    const val SLOTS_PER_BEAT = 4   // sixteenth grid

    val MIN_BARS = 1
    val MAX_BARS = 4
    val TIME_SIGNATURES = listOf(2, 3, 4)   // N/4

    /** Units a phrase can be built from: the 16th-grid units (no triplet) + rests. */
    val POOL: List<RhythmUnit> = RhythmUnits.ALL.filter { it.subdivision == SLOTS_PER_BEAT } + RhythmUnits.RESTS

    fun generatePhrase(bars: Int, beatsPerBar: Int, rng: kotlin.random.Random): RhythmPhrase {
        val b = bars.coerceIn(MIN_BARS, MAX_BARS)
        val bpb = if (beatsPerBar in TIME_SIGNATURES) beatsPerBar else 2
        val beats = List(b * bpb) { POOL[rng.nextInt(POOL.size)] }
        return RhythmPhrase(b, bpb, beats)
    }
}

File: chorect-web/src/theory/rhythmPhrases.ts
// Multi-bar rhythmic phrases. Mirror of theory/.../RhythmPhrases.kt.

import { RhythmUnit, RHYTHM_UNITS, RHYTHM_UNITS_RESTS, clickFractions } from "./rhythmUnits";
import { Rng, defaultRng } from "./random";

export const SLOTS_PER_BEAT = 4;   // sixteenth grid
export const PHRASE_MIN_BARS = 1;
export const PHRASE_MAX_BARS = 4;
export const PHRASE_TIME_SIGNATURES = [2, 3, 4];   // N/4

/** Units a phrase can be built from: the 16th-grid units (no triplet) + rests. */
export const PHRASE_POOL: RhythmUnit[] =
  [...RHYTHM_UNITS.filter((u) => u.subdivision === SLOTS_PER_BEAT), ...RHYTHM_UNITS_RESTS];

export interface PhraseOnset { slot: number; accent: boolean; }

export interface RhythmPhrase {
  bars: number;
  beatsPerBar: number;
  beats: RhythmUnit[];   // length = bars*beatsPerBar, each subdivision 4
}

export function phraseTotalSlots(p: RhythmPhrase): number { return p.beats.length * SLOTS_PER_BEAT; }
export function phraseSlotsPerBar(p: RhythmPhrase): number { return p.beatsPerBar * SLOTS_PER_BEAT; }

/** Global 16th-slot of every click, accented on bar downbeats that carry an onset. */
export function phraseOnsets(p: RhythmPhrase): PhraseOnset[] {
  const out: PhraseOnset[] = [];
  p.beats.forEach((unit, b) => {
    const isBarStart = b % p.beatsPerBar === 0;
    for (const f of clickFractions(unit)) {
      const local = Math.round(f * SLOTS_PER_BEAT);
      out.push({ slot: b * SLOTS_PER_BEAT + local, accent: isBarStart && local === 0 });
    }
  });
  return out;
}

export function generatePhrase(bars: number, beatsPerBar: number, rng: Rng = defaultRng): RhythmPhrase {
  const b = Math.min(Math.max(bars, PHRASE_MIN_BARS), PHRASE_MAX_BARS);
  const bpb = PHRASE_TIME_SIGNATURES.includes(beatsPerBar) ? beatsPerBar : 2;
  const beats = Array.from({ length: b * bpb }, () => PHRASE_POOL[rng.int(PHRASE_POOL.length)]);
  return { bars: b, beatsPerBar: bpb, beats };
}
